mirgecom/artificial_viscosity.py

```diff
@@ -0,0 +1,328 @@
+    else:
+        actx = flux_dis[0].array_context
+
+    normal = thaw(actx, discr.normal(r_tpair.dd))
+
+    # This uses a central vector flux along nhat:
+    # flux = 1/2 * (grad(Q)- + grad(Q)+) .dot. nhat
+    flux_out = flux_dis @ normal
+
+    return discr.project(r_tpair.dd, "all_faces", flux_out)
+
+
+def av_operator(discr, t, eos, boundaries, q, alpha, **kwargs):
+    r"""Compute the artificial viscosity right-hand-side.
+
+    Computes the the right-hand-side term for artificial viscosity.
+
+    .. math::
+
+        \mbox{RHS}_{\mbox{av}} = \nabla\cdot{\varepsilon\nabla\mathbf{Q}}
+
+    Parameters
+    ----------
+    q: :class:`~meshmode.dof_array.DOFArray` or :class:`~numpy.ndarray`
+
+        Single :class:`~meshmode.dof_array.DOFArray` for a scalar or an object array
+        (:class:`~numpy.ndarray`) for a vector of
+        :class:`~meshmode.dof_array.DOFArray` on which to operate.
+
+        When used with fluid solvers, *q* is expected to be the fluid state array
+        of the canonical conserved variables (mass, energy, momentum)
+        for the fluid along with a vector of species masses for multi-component
+        fluids.
+
+    boundaries: float
+
+        Dictionary of boundary functions, one for each valid boundary tag
+
+    t: float
+
+        Time
+
+    alpha: float
+
+       The maximum artificial viscosity coefficient to be applied
+
+    eos: :class:`~mirgecom.eos.GasEOS`
+
+       Only used as a pass through to the boundary conditions.
+
+    Returns
+    -------
+    numpy.ndarray
+
+        The artificial viscosity operator applied to *q*.
+    """
+    # Get smoothness indicator based on first component
+    indicator_field = q[0] if isinstance(q, np.ndarray) else q
+    indicator = smoothness_indicator(discr, indicator_field, **kwargs)
+
+    # R=Grad(Q) volume part
+    if isinstance(q, np.ndarray):
+        grad_q_vol = np.stack(obj_array_vectorize(discr.weak_grad, q), axis=0)
+    else:
+        grad_q_vol = discr.weak_grad(q)
+
+    # R=Grad(Q) Q flux over interior faces
+    q_flux_int = _facial_flux_q(discr, q_tpair=interior_trace_pair(discr, q))
+    # R=Grad(Q) Q flux interior faces on partition boundaries
+    q_flux_pb = sum(_facial_flux_q(discr, q_tpair=pb_tpair)
+                    for pb_tpair in cross_rank_trace_pairs(discr, q))
+    # R=Grad(Q) Q flux domain boundary part (i.e. BCs)
+    q_flux_db = 0
+    for btag in boundaries:
+        q_tpair = TracePair(
+            btag,
+            interior=discr.project("vol", btag, q),
+            exterior=boundaries[btag].exterior_q(discr, btag=btag, t=t,
+                                                 q=q, eos=eos))
+        q_flux_db = q_flux_db + _facial_flux_q(discr, q_tpair=q_tpair)
+    # Total Q flux across element boundaries
+    q_bnd_flux = q_flux_int + q_flux_pb + q_flux_db
+
+    # Compute R
+    r = discr.inverse_mass(
+        -alpha * indicator * (grad_q_vol - discr.face_mass(q_bnd_flux))
```

Comment:
When I did related work, I found it essential to use a "smooth" viscosity (otherwise strong AV might produce more nonsmoothness than it would remove). I got acceptable results using the following steps:
![image](https://user-images.githubusercontent.com/352067/116784403-9d221580-aa59-11eb-8928-f496167054c1.png)

(https://doi.org/10.1051/mmnp/20116303, Section 5.2)